### PR TITLE
NetworkController: Match `rollbackToPreviousProvider` to extension

### LINKED ISSUE #1220: NetworkController: `rollbackToPreviousProvider` does not account for RPC-based provider config without associated network configuration ID
If you initialize NetworkController with the following in state:

```
{
  providerConfig: {
    type: NetworkType.rpc,
    rpcTarget: 'https://mock-rpc-url',
    chainId: '0x1337',
  }
}
```

then if you switch the network, then call `rollbackToPreviousProvider` again, the result won't be what you expect.

Facts:

* This version of NetworkController has two methods to switch networks: `setProviderType` and `setActiveNetwork`. (The latter method replaced `setRpcTarget` when network configurations were introduced.)

* `rollbackToPreviousProvider` checks to see what `#previousNetworkSpecifier` is to figure out whether to call `setProviderType` or `setActiveNetwork`.

* What is `#previousNetworkSpecifier`? This property is designed to hold either a network configuration ID, or a "provider type" (i.e. anything that `#configureProvider` recognizes). But it will only hold a network configuration ID if it refers to a custom endpoint; otherwise it holds the provider type.

* When does `#previousNetworkSpecifier` get set? In two instances:

  1. Before the network gets switched. In this case if the old network is a custom network (provider type is "rpc") and the provider config has a reference to a network configuration ID, then it'll be that network configuration ID. Otherwise, it'll be the provider type.
  2. When the NetworkConfiguration is initialized. In this case it's always the provider type, regardless of whether it's "rpc" or not.
  
So what does this mean? Because `#previousNetworkSpecifier` is set in the constructor and is always set to the provider type, that means that it is possible for `#previousNetworkSpecifier` to be "rpc" and the provider config have no reference to a network configuration ID. In that case, `rollbackToPreviousProvider` will try to call `setProviderType` with a provider type of "rpc". This doesn't make any sense, because it's designed to be passed an _Infura_ provider type. Because in that case the `rpcTarget` is cleared out, attempting to talk to the network after that wouldn't work.

### PR DESCRIPTION
**Pull request by mcmire**
## Description

When `rollbackToPreviousProvider` is called, it will use
`#previousNetworkSpecifier` to determine how to switch the network back
to the previous one. This property holds one of two values depending on
how that network was set in the first place:

- If `setActiveNetwork` was used to set the network, then
  `#previousNetworkSpecifier` will be the ID of a network configuration.
- If `setProviderType` was used to set the network, then
  `#previousNetworkSpecifier` will be the `type` in the provider config.
- If the network controller has just been initialized, then if the
  `type` in the provider config is "rpc" and the provider config holds a
  reference to a network configuration, then `#previousNetworkSpecifier`
  will be its ID, otherwise it will be whatever the provider `type` is.

This means that when `rollbackToPreviousProvider` is called, it can ask
what `#previousNetworkSpecifier` is and use its value to route to the
correct method:

- If `#previousNetworkSpecifier` is a known provider type, then call
  `setProviderType`, passing the type.
- Otherwise, if `#previousNetworkSpecifier` is a string, assume that it
  is a network configuration ID and call `setActiveNetwork` with it.

However, a problem arises when the network has not been switched yet,
and the provider config has a `type` of "rpc" but does not contain a
reference to a network configuration. Because
`#previousNetworkSpecifier` in that case is a provider type,
`rollbackToPreviousProvider` would try to call `setProviderType` with
"rpc". `setProviderType` is meant to be passed an _Infura_-specific
provider type, i.e. the name of an Infura network, so it will throw if
it doesn't receive one.

There are probably a couple of different ways to solve this problem,
but we can consider what the extension version of NetworkController
does: when the controller is initialized or before the network is
switched, the current provider config is saved, the whole object is
loaded into `providerConfig` in state when `rollbackToPreviousProvider`
is called, and then `setProviderType` and `setActiveNetwork` are
completely skipped.

Because that implementation already works, it has been copied into this
version of NetworkController. The tests for `rollbackToPreviousProvider`
have also been updated to match the extension as well.

## Changes

- **FIXED**: Fix `rollbackToPreviousProvider` in NetworkController so that it does not throw if the `type` in provider config is "rpc" but has no reference to a network configuration.

## References

Fixes #1220.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation for new or updated code as appropriate (note: this will usually be JSDoc)
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/network-controller/tests/NetworkController.test.ts [hunk 1]
```diff
@@ -5297,764 +5333,885 @@ describe('NetworkController', () => {
...
-            }),
-          };
-
           await withController(
             {
               state: {
-                providerConfig: initialProviderConfig,
+                providerConfig: {
+                  type: networkType,
+                  // NOTE: This doesn't need to match the logical chain ID
```
Comment (Gudahtt) [inline packages/network-controller/tests/NetworkController.test.ts:5344]: Nit: Maybe better to use `buildProviderConfig({ ...BUILT_IN_NETWORKS[networkType], type: networkType })` here instead, to keep this more concise and avoid cluttering up the test too much.

All of these extra details makes it harder to tell what condition is being tested, and using an unrealistic initial state (though still possible from an API standpoint) makes the test less similar to production conditions.

File: packages/network-controller/tests/NetworkController.test.ts [hunk 2]
```diff
@@ -5297,764 +5333,885 @@ describe('NetworkController', () => {
...
+                nickname: 'test network 1',
+                rpcPrefs: {
+                  blockExplorerUrl: 'https://test-block-explorer-1.com',
+                },
+              });
             },
           );
         });
 
-        it('emits NetworkController:providerConfigChange via the messenger', async () => {
```
Comment (Gudahtt) [inline packages/network-controller/tests/NetworkController.test.ts:5364]: Nit: it seems that this was removed accidentally? (and same for the RPC version of this test). But we're about to remove it anyway, so it's not such a big deal

Comment (mcmire) [reply packages/network-controller/tests/NetworkController.test.ts:4609]: I'll add it back for completeness, for now, until we remove it.

File: packages/network-controller/tests/NetworkController.test.ts [hunk 3]
```diff
@@ -6500,10 +6721,16 @@ async function waitForPublishedEvents<E extends NetworkControllerEvents>(
...
-      function end() {
+      async function end() {
         if (!alreadyEnded) {
           messenger.unsubscribe(eventType, eventListener);
-          if (interestingEventPayloads.length === expectedNumberOfEvents) {
+
+          await beforeResolving();
+
+          if (
+            expectedNumberOfEvents === null ||
```
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/network-controller/tests/NetworkController.test.ts]: Nope! Removed in [2f292aa](https://github.com/MetaMask/core/pull/1223/commits/2f292aa868bf74afbfd3011eec70901fa93e7b1b).

File: packages/network-controller/tests/NetworkController.test.ts (hunk 1 above)
Comment (mcmire) [reply packages/network-controller/tests/NetworkController.test.ts]: Good point. Fixed in [f3b02f7](https://github.com/MetaMask/core/pull/1223/commits/f3b02f77a950ed909aab5f835e1bde0af32e1cba).

**Comment by mcmire** [comment]
@Gudahtt I believe I've addressed all of your feedback.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Comment by mcmire** [comment]
Fixed conflicts.

**Review by Gudahtt — DISMISSED** [review]

**Review by Gudahtt — APPROVED** [review]

**Pull request merged**